typing/env.ml

```diff
@@ -221,8 +221,8 @@ module IdTbl =
 
 
     type ('a, 'b) t = {
-      current: 'a Ident.tbl;
-      (** Local bindings since the last open *)
+      current: (alloc_mode * 'a) Ident.tbl;
```

Comment:
I thought about this a bit, and I think the cleanest option is to make alloc_mode be a type parameter here. (I don't fully follow the suggestion about `Map`).

Reply:
`Map` let's you add an arbitrary function that is mapped on all values the are looked up through that `Map`. This is pretty similar to the definition of a lock. If you have a lock as part of `value_entry` (which would be just an "empty" lock initially) then you can use something like:
```ocaml
IdTbl.map (function
  | Val_unbound _ as desc -> desc
  | Val_bound(desc, lock) -> Val_bound(desc, add_lock new_lock lock)) env
```
to add `new_lock` to `env`.

Reply:
You most recent version looks pretty good. But I think it would be simpler if the type variable represented just the locks not the locks and the original mode. I think you could then push the original mode into `value_entry` or `value_data`.